### PR TITLE
ipld: retrieve data starting from root

### LINKED ISSUE #247: Improve RetrieveBlockData
## Summary
Current RetrieveBlockData implementation spawns a goroutine per each share in ExtendedDataSquare. In the worst case with the max block 128X128 that is 16384 goroutines. This causes the race detector to stop working in tests, as it has a limit to 8k routines. Furthermore, this a lot of routines for a single operation, and in fact most of the time they are idle. 

Why routines are idle. Shares in a block are addressed and committed with multiple Merkle Trees, so when we request a whole block we need to walk through all the trees. The walking step is: request blob of data by its hash(1), unpack block(2), check if we unpacked more hashes or a share(3), and proceed with unpacked hashes onto the next step recursively(3). Those steps are executed until no more hashes left and we have all the shares. From this, we can see that every walking step is a full roundtrip - network request, unpack, request again, and so on. Now, imagine those 16384 goroutines which walk down the same trees. In practice, those routines mostly wait for every single roundtrip to finish and then they compete to initiate the next roundtrip. Competing, in this case, is useless and we need to avoid it.

Instead of spawning goroutine per share, we should spawn a routine per tree we need to walk through, particularly per every DAHeader root. This way, we don't have any competing routines and only one routine initiates all ther roundtrips. In numbers, for the largest block that would be 128 routines per Row DAHeader root and 256 routines if want to fetch and store all the inner nodes of trees. 
NOTE: All the shares are addressed and committed twice, with two Row and Column Merkle Trees, thus it is not required to traverse both trees to fetch the shares, but again, if we need to store all inner nodes of both trees(cc @adlerjohn) than both tree should be traversed.

## Implementation Details
// TODO

### Action Items
// TODO

### References
As already mentioned, the original spark of the issue is the race detector complaining about exceeding the limit of 8k routines. The solution for that was to detect if the race detector running and simply skipping the test :grimacing:. Later, after some discussions, https://github.com/celestiaorg/lazyledger-core/issues/357 was created. And the solution for it is being implemented here https://github.com/celestiaorg/lazyledger-core/pull/424.

Also, supersede https://github.com/celestiaorg/lazyledger-core/issues/278

## Appendix
There is a plan to remove DAHeader and use only one root hash to address and commit to all the shares. In such a case, we would not need to spawn any routines at all. The routine calling the RetrieveBlockData needs to be blocking anyway and it can then take care of initiating all the round trips.

### PR DESCRIPTION
**Pull request by vgonkivs**
Fixes #247

* fetching data from root instead of particular share;
* removed share counter;

Time duration from start fetching till the end of fetching:
```
=== RUN   TestRetrieveBlockData
=== RUN   TestRetrieveBlockData/1x1(min)
374.436µs
=== RUN   TestRetrieveBlockData/32x32(med)
52.936894ms
=== RUN   TestRetrieveBlockData/128x128(max)
1.298028147s
```

Unfortunately, test  TestRetrieveBlockData/128x128(max) failed by timeout. I guess bottleneck could be [here](https://github.com/celestiaorg/rsmt2d/blob/ac0f1e1a51bf7b5420965fb7c35fa32a56e02292/extendeddatacrossword.go#L77). In the worst case we are  iterating over a slice of 65k of elements and making a pretty heavy operation of copying.

### CODE REVIEW

File: ipld/read.go [hunk 1]
```diff
@@ -20,188 +18,130 @@ import (
...
-		cancel:          cancel,
+			leafs, err := GetLeafs(ctx, dag, *subtreeRootHash)
+			if err != nil {
+				f.err <- err
+				return
+			}
+			for leafIdx, leaf := range leafs {
+				shareData := leaf.RawData()[1:]
+				// it's not needed to store data for cols
+				// as we are fetching data from the same share for rows and cols
```
Comment (renaynay) [inline ipld/read.go:91]: Can you also add a document as to why we still fetch cols even though we're not storing them in the data slice? cc @Wondertan

File: ipld/read.go [hunk 2]
```diff
@@ -20,188 +18,130 @@ import (
...
-func (sc *shareCounter) retrieveShare(
-	rootCid cid.Cid,
-	isRow bool,
-	axisIdx uint32,
-	idx uint32,
-	dag ipld.NodeGetter,
-) {
-	data, err := GetLeafData(sc.ctx, rootCid, idx, sc.edsWidth, dag)
+// getHalfTreeNodeHash returns only one subtree - left or right
+func getHalfTreeNodeHash(ctx context.Context, root cid.Cid, dag ipld.NodeGetter, isLeftSubtree bool) (*cid.Cid, error) {
```
Comment (renaynay) [inline ipld/read.go:102]: You're always calling this with `isLeftSubtree` as true, so why not just have this method return the left subtree only instead of passing in the bool?

File: ipld/read.go [hunk 3]
```diff
@@ -20,188 +18,130 @@ import (
...
-			if !has {
-				sc.shares[share.index] = share.data
-				sc.counter++
-				// check finishing condition
-				if sc.counter >= sc.minSharesNeeded {
-					return nil
-				}
-			}
+// GetLeafs recursively starts going down to find all leafs from the given root
+func GetLeafs(ctx context.Context, dag ipld.NodeGetter, root cid.Cid) (leafs []ipld.Node, err error) {
```
Comment (renaynay) [inline ipld/read.go:116]: Please rename to `GetLeaves`

**Review by renaynay — COMMENTED** [review]
Initial review - left some comments but looks good in general! Thank you for your work on this :)

File: ipld/read.go (hunk 2 above)
Comment (Wondertan) [inline ipld/read.go:102]: Cid's are meant to be copied and passed as values, even it is a struct. Not sure why you kept root CID as value and the returning one as a pointer.

Comment (Wondertan) [inline ipld/read.go:102]: Let's make this func public and call it `GetSubtreeLeaves` so it also returns the actual leaves of subtree by calling `GetLeaves` internally. This will make the code cleaner.

File: ipld/read.go [hunk 4]
```diff
@@ -20,188 +18,130 @@ import (
...
+			}
+			for leafIdx, leaf := range leafs {
+				shareData := leaf.RawData()[1:]
+				// it's not needed to store data for cols
+				// as we are fetching data from the same share for rows and cols
+				if isRow {
+					f.data[(i*len(data))+leafIdx] = shareData[NamespaceSize:]
+				}
+			}
+			f.err <- nil
```
Comment (Wondertan) [inline ipld/read.go:96]: This could be a func that returns an error instead of writing to chan in each err != nil

File: ipld/read.go [hunk 5]
```diff
@@ -20,188 +18,130 @@ import (
...
+		select {
+		case err := <-shares.err:
 			if err != nil {
 				return nil, err
 			}
-
-			go sc.retrieveShare(rootCid, true, row, col, dag)
+		case <-ctx.Done():
+			return nil, ctx.Err()
 		}
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply ipld/read.go:56]: This will remove the need to make a custom totalShares types. Let's pls avoid spamming types if we don't have to, so we keep things allocated on the stack and not the heap.

**Review by Wondertan — CHANGES_REQUESTED** [review]
Initial review. LGTM. Learning quick

File: ipld/read.go [hunk 6]
```diff
@@ -20,188 +18,130 @@ import (
...
-// too many errors are encountered
-func (sc *shareCounter) wait() error {
-	defer sc.cancel()
-
-	for {
-		select {
-		case <-sc.ctx.Done():
-			err := sc.ctx.Err()
-			if err == context.DeadlineExceeded {
-				return ErrRetrieveTimeout
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply ipld/read.go:178]: removed

File: ipld/read.go (hunk 3 above)
Comment (vgonkivs) [reply ipld/read.go]: Renamed

File: ipld/read.go [hunk 7]
```diff
@@ -20,188 +18,130 @@ import (
...
+type totalShares struct {
+	data [][]byte
+	err  chan error
+}
+
 // RetrieveData asynchronously fetches block data using the minimum number
 // of requests to IPFS. It fails if one of the random samples sampled is not available.
 func RetrieveData(
-	ctx context.Context,
+	parentCtx context.Context,
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply ipld/read.go]: parentCtx removed

File: ipld/read.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (vgonkivs) [reply ipld/read.go]: function reworked

File: ipld/read.go (hunk 2 above)
Comment (vgonkivs) [reply ipld/read.go]: changed pointer to value in return

Comment (vgonkivs) [reply ipld/read.go]: renamed

File: ipld/read.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vgonkivs) [reply ipld/read.go]: reworked to be able to work with errgroup.

File: ipld/read.go (hunk 4 above)
Comment (vgonkivs) [reply ipld/read.go]: reworked

File: ipld/read.go [hunk 8]
```diff
@@ -29,179 +25,117 @@ func RetrieveData(
...
-	// wait until enough data has been collected, too many errors encountered,
-	// or the timeout is reached
-	err := sc.wait()
-	if err != nil {
+	errGroup, ctx := errgroup.WithContext(ctx)
+	errGroup.Go(func() error {
+		return fillQuarter(ctx, rowRoots, dag, true, dataSquare)
+	})
+	errGroup.Go(func() error {
+		return fillQuarter(ctx, rowRoots, dag, false, dataSquare)
```
Comment (liamsi) [inline ipld/read.go:37]: Are these supposed to be row roots in both cases?

File: ipld/read.go [hunk 9]
```diff
@@ -29,179 +25,117 @@ func RetrieveData(
...
-	errc chan error
+	return rsmt2d.RepairExtendedDataSquare(rowRoots, colRoots, dataSquare, codec, tree.Constructor)
 }
 
-func newshareCounter(parentCtx context.Context, edsWidth uint32) *shareCounter {
-	ctx, cancel := context.WithCancel(parentCtx)
+// fillQuarter fetches 1/4 of shares for the given root
+func fillQuarter(
+	ctx context.Context,
+	data [][]byte, dag ipld.NodeGetter,
```
Comment (liamsi) [inline ipld/read.go:50]: Shouldn't this be called `roots` instead of just `data`.
Also why has ever param its own line while dad and data share one?

File: ipld/read.go (hunk 8 above)
Comment (vgonkivs) [reply ipld/read.go]: Ah, yeah. Nice catch. Thank you. Changed to cols for the second case.

File: ipld/read.go (hunk 9 above)
Comment (vgonkivs) [reply ipld/read.go]: Agree. Renamed to roots.

File: ipld/read.go (hunk 2 above)
Comment (renaynay) [reply ipld/read.go]: Can we simplify it for this PR at least to only have this function do `getLeftHalfTreeNodeHash` ? It's cleaner if this is the only usage *for now* IMO.

**Review by renaynay — APPROVED** [review]
LGTM now that @Wondertan 's comments have been addressed.

**Comment by renaynay** [comment]
Needs a rebase though

**Comment by renaynay** [comment]
Unfortunately needs another rebase :(

**Comment by vgonkivs** [comment]
Rebased

File: service/share/full_availability.go [hunk 10]
```diff
@@ -13,19 +13,19 @@ import (
 // recovery technique. It is considered "full" because it is required
 // to download enough shares to fully reconstruct the data square.
 type fullAvailability struct {
-	getter format.NodeGetter
+	service format.DAGService
```
Comment (liamsi) [inline service/share/full_availability.go:16]: Just a question: what triggered changing this to point to a DAGService instead of a NodeGetter? From reading the changes it seems that everywhere only the the NodeGetter part is used. Or am I missing sth?

**Review by liamsi — APPROVED** [review]
👏🏼

File: CHANGELOG-PENDING.md [hunk 11]
```diff
@@ -7,7 +7,7 @@
 ### FEATURES
 
 ### IMPROVEMENTS
-
+- [ipld: retreive data starting from root #442](https://github.com/celestiaorg/celestia-node/pull/442) [@vgonkivs](https://github.com/vgonkivs)
```
Comment (adlerjohn) [inline CHANGELOG-PENDING.md:10]:
```suggestion

- [ipld: retrieve data starting from root #442](https://github.com/celestiaorg/celestia-node/pull/442) [@vgonkivs](https://github.com/vgonkivs)

```

File: ipld/read.go [hunk 12]
```diff
@@ -18,190 +16,131 @@ import (
...
-			// add iff it does not already exists
-			if !has {
-				sc.shares[share.index] = share.data
-				sc.counter++
-				// check finishing condition
-				if sc.counter >= sc.minSharesNeeded {
-					return nil
-				}
-			}
+// getLeaves recursively starts going down to find all leafs from the given root
```
Comment (adlerjohn) [inline ipld/read.go:116]: What is the difference between `GetLeaves` and `getLeaves`? Looks like the latter is the implementation? They should probably have different names.

File: service/share/full_availability.go (hunk 10 above)
Comment (vgonkivs) [reply service/share/full_availability.go:16]: Because now, we are committing shares to ipfs after successful repairing. And it requires ipld.NodeAdder.

File: ipld/read.go (hunk 12 above)
Comment (vgonkivs) [reply ipld/read.go:116]: `GetLeaves` is a wrapper over `getLeaves` but I think they both have correct namings. I've decided to move more complex logic to a separate function.

**Review by renaynay — APPROVED** [review]

File: ipld/read.go [hunk 13]
```diff
@@ -18,190 +16,131 @@ import (
...
-	errc chan error
-}
-
-func newshareCounter(parentCtx context.Context, edsWidth uint32) *shareCounter {
-	ctx, cancel := context.WithCancel(parentCtx)
+// fillQuarter fetches 1/4 of shares for the given root
+func fillQuarter(
+	ctx context.Context,
+	roots [][]byte, dag ipld.NodeGetter,
+	isRow bool,
```
Comment (adlerjohn) [inline ipld/read.go:56]: What's the purpose of `isRow`? It looks like you only ever set it to `true`...

File: ipld/read.go [hunk 14]
```diff
@@ -18,190 +16,131 @@ import (
...
-	ctx       context.Context
-	cancel    context.CancelFunc
-	// any errors encountered when attempting to retrieve shares
-	errc chan error
-}
-
-func newshareCounter(parentCtx context.Context, edsWidth uint32) *shareCounter {
-	ctx, cancel := context.WithCancel(parentCtx)
+// fillQuarter fetches 1/4 of shares for the given root
+func fillQuarter(
```
Comment (adlerjohn) [inline ipld/read.go:53]: Nit, but I would prefer `Quadrant` instead of `Quarter`

Comment (Wondertan) [reply ipld/read.go:53]: Agree with John

**Comment by Wondertan** [comment]
@vgonkivs, will review after quadrant randomization is implemented.

File: ipld/read.go [hunk 15]
```diff
@@ -18,190 +18,162 @@ import (
...
+	return extended, batchAdder.Commit()
 }
 
-type index struct {
-	row uint32
-	col uint32
+type quadrant struct {
+	isLeftSubtree bool
+	from          int
+	chunk         [][]byte //roots
```
Comment (Wondertan) [inline ipld/read.go:53]: Rename to Root. Also, I suggest having roots be always as CID's so you convert them in the beginning. This way it's also clear what roots are and not just arbitrary bytes.

File: ipld/read.go [hunk 16]
```diff
@@ -18,190 +18,162 @@ import (
...
-	// number of shares successfully collected
-	counter uint32
-	// the width of the extended data square
-	edsWidth uint32
-	// the minimum shares needed to repair the extended data square
-	minSharesNeeded uint32
+	// define range of shares for sampling
+	if q > 2 {
+		quadrant.from = edsWidth / 2
+		copy(quadrant.chunk, roots[edsWidth/2:])
```
Comment (Wondertan) [inline ipld/read.go]: You could also point to the full root slice in the code where you use roots, instead of copying the half.

File: ipld/read.go [hunk 17]
```diff
@@ -18,190 +18,162 @@ import (
...
+				return err
+			}
+			subtreeRootHash, err := GetSubtreeLeaves(ctx, rootHash, dag, roots.isLeftSubtree)
+			if err != nil {
+				return err
+			}
 
-	// calculate the min number of shares needed to repair the square
-	minSharesNeeded := edsWidth * edsWidth / 4
+			leaves, err := GetLeaves(ctx, dag, subtreeRootHash, uint32(len(roots.chunk)))
```
Comment (Wondertan) [inline ipld/read.go:101]: Should be the size of roots divided by two

File: ipld/read.go (hunk 1 above)
Comment (Wondertan) [reply ipld/read.go:114]: As per multiple discussionss, no need to fetch cols, but only as retry if fails for rows

File: ipld/read.go [hunk 18]
```diff
@@ -18,190 +18,162 @@ import (
...
-func (sc *shareCounter) retrieveShare(
-	rootCid cid.Cid,
-	isRow bool,
-	axisIdx uint32,
-	idx uint32,
-	dag ipld.NodeGetter,
-) {
-	data, err := GetLeafData(sc.ctx, rootCid, idx, sc.edsWidth, dag)
+// GetSubtreeLeaves returns only one subtree - left or right
+func GetSubtreeLeaves(ctx context.Context, root cid.Cid, dag ipld.NodeGetter, isLeftSubtree bool) (cid.Cid, error) {
```
Comment (Wondertan) [inline ipld/read.go:130]: Can you make this func to return the actual subtree by calling GetLeaves internally?

File: ipld/read.go [hunk 19]
```diff
@@ -18,190 +17,166 @@ import (
...
-}
+// fillQuadrant fetches 1/4 of shares for the given root
+func fillQuadrant(
+	ctx context.Context,
+	roots *quadrant, dag ipld.NodeGetter,
+	isRow bool,
+	dataSquare [][]byte,
+) error {
+	errGroup, ctx := errgroup.WithContext(ctx)
+	fetcher := func(i int) {
```
Comment (Wondertan) [inline ipld/read.go:90]: You can avoid defining the func by doing `i := i`

File: ipld/read.go [hunk 20]
```diff
@@ -18,190 +17,166 @@ import (
...
-	ctx       context.Context
-	cancel    context.CancelFunc
-	// any errors encountered when attempting to retrieve shares
-	errc chan error
-}
+// fillQuadrant fetches 1/4 of shares for the given root
+func fillQuadrant(
+	ctx context.Context,
+	roots *quadrant, dag ipld.NodeGetter,
+	isRow bool,
```
Comment (Wondertan) [inline ipld/read.go:86]: Delete isRow

File: ipld/read.go (hunk 13 above)
Comment (vgonkivs) [reply ipld/read.go:86]: But then we will add it again in scope of implementing retry mechanism from #508

File: ipld/read.go [hunk 21]
```diff
@@ -18,190 +17,166 @@ import (
...
+				// shares from the right subtree should be
+				// inserted after all shares from the left subtree
+				if !roots.isLeftSubtree {
+					leafIdx += len(leaves)
+				}
+				shareData := leaf.RawData()[1:]
+				// it's not needed to store data for cols
+				// as we are fetching data from the same share for rows and cols
+				if isRow {
+					dataSquare[((i+roots.from)*len(roots.chunk)*2)+leafIdx] = shareData[NamespaceSize:]
```
Comment (Wondertan) [inline ipld/read.go:116]: `len(roots.chunk)` does not change so can be called once in the func

File: ipld/read.go (hunk 16 above)
Comment (Wondertan) [reply ipld/read.go]: But that is an optional nit

File: ipld/read.go (hunk 14 above)
Comment (vgonkivs) [reply ipld/read.go]: Renamed

File: ipld/read.go (hunk 13 above)
Comment (vgonkivs) [reply ipld/read.go]: removed

File: ipld/read.go (hunk 15 above)
Comment (vgonkivs) [reply ipld/read.go]: renamed to `rootCids`

File: ipld/read.go (hunk 16 above)
Comment (vgonkivs) [reply ipld/read.go]: changed type to `cid.Cid`. So, there is no needs of copying anymore.

File: ipld/read.go (hunk 17 above)
Comment (vgonkivs) [reply ipld/read.go]: done

File: ipld/read.go (hunk 18 above)
Comment (vgonkivs) [reply ipld/read.go]: moved to `GetSubtreeLeaves`

File: ipld/read.go (hunk 19 above)
Comment (vgonkivs) [reply ipld/read.go]: done

File: ipld/read.go (hunk 20 above)
Comment (vgonkivs) [reply ipld/read.go]: removed

File: ipld/read.go (hunk 21 above)
Comment (vgonkivs) [reply ipld/read.go]: done

File: ipld/read.go [hunk 22]
```diff
@@ -18,190 +17,170 @@ import (
...
+	// quadrants 1 and 3 corresponds to left subtree,
+	// 2 and 4 to the right subtree
+	/* | 1 | 2 |
+	   | 3 | 4 |
+	*/
+	// choose subtree
+	if q%2 == 1 {
+		quadrant.isLeftSubtree = true
+	}
+	to := 0
```
Comment (renaynay) [inline ipld/read.go:73]: I don't understand why `from` is a field inside of `quadrant` but `to` is not, and vice versa. Is there a reason?

File: ipld/read.go [hunk 23]
```diff
@@ -18,190 +17,170 @@ import (
...
-	// any errors encountered when attempting to retrieve shares
-	errc chan error
+	var err error
+	for index, counter := quadrant.from, 0; index < to; index++ {
+		quadrant.rootCids[counter], err = plugin.CidFromNamespacedSha256(roots[index])
+		if err != nil {
+			return nil, err
+		}
+		counter++
+	}
```
Comment (renaynay) [inline ipld/read.go:89]: if both `counter` and `index` are incrementing +1, why not just use `index`?

File: ipld/read.go [hunk 24]
```diff
@@ -18,190 +17,170 @@ import (
...
-	axisIdx uint32,
-	idx uint32,
+// GetSubtreeLeaves returns only one subtree - left or right
+func GetSubtreeLeaves(
+	ctx context.Context,
+	root cid.Cid,
 	dag ipld.NodeGetter,
-) {
-	data, err := GetLeafData(sc.ctx, rootCid, idx, sc.edsWidth, dag)
+	isLeftSubtree bool, treeSize uint32) ([]ipld.Node, error) {
```
Comment (renaynay) [inline ipld/read.go:129]:
```suggestion
	isLeftSubtree bool, 
	treeSize uint32,
	) ([]ipld.Node, error) {
```

File: ipld/read.go [hunk 25]
```diff
@@ -18,190 +17,170 @@ import (
...
+}
 
-	select {
-	case <-sc.ctx.Done():
-	case sc.shareChan <- indexedShare{data: data[NamespaceSize:], index: index{row: rowIdx, col: colIdx}}:
+func GetLeaves(ctx context.Context, dag ipld.NodeGetter, root cid.Cid, size uint32) ([]ipld.Node, error) {
+	leaves, err := getLeaves(ctx, dag, root, make([]ipld.Node, 0, size))
+	if err != nil {
+		return nil, err
 	}
```
Comment (renaynay) [inline ipld/read.go:151]: I think you can just return `getLeaves`, etc.

File: ipld/read.go (hunk 23 above)
Comment (vgonkivs) [reply ipld/read.go:89]: because `counter` and `index` could not be equal.

Comment (renaynay) [reply ipld/read.go:89]: Oh sorry, I thought I saw 
```
for index, counter := quadrant.from
```

but it's actually

```
for index, counter := quadrant.from, 0
```

**Review by Wondertan — APPROVED** [review]
After @renaynay comments are applied this is fine to me. `pickRandomQuadrant` still feels a bit ugly to me, but I cant propose anything better atm and this PR is better to be merged so :shipit:

File: ipld/read.go (hunk 22 above)
Comment (vgonkivs) [reply ipld/read.go:73]: We need `from` field in fillQuadrant to correctly set index of share. But `to`  var is used to define the last root that we need to store from RowRoots, since the quadrant can have two different ranges:
* [0;edsWidth/2)
* [edsWidth/2;edsWidth)

Comment (renaynay) [reply ipld/read.go:73]: I still don't understand why `to` can't also be a field set within `quadrant` ? it's just a bit hard to follow the code here.

File: ipld/read.go [hunk 26]
```diff
@@ -18,190 +17,170 @@ import (
...
-	ctx       context.Context
-	cancel    context.CancelFunc
-	// any errors encountered when attempting to retrieve shares
-	errc chan error
+	var err error
+	for index, counter := quadrant.from, 0; index < to; index++ {
+		quadrant.rootCids[counter], err = plugin.CidFromNamespacedSha256(roots[index])
+		if err != nil {
+			return nil, err
+		}
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply ipld/read.go:87]: In the beginning of the RetrieveData we don't know what quadrant will we choose and thus generate cids for each root in row/col. Do we need this?

File: ipld/read.go (hunk 22 above)
Comment (vgonkivs) [reply ipld/read.go]: extended `quadrant` struct with `to` field

Comment (renaynay) [reply ipld/read.go]: But neither does `quadrant.from` right?

File: ipld/read.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (renaynay) [reply ipld/read.go:89]: I'd recommend making an issue out of this discussion, merging as is and then coming back to it later, if it's important.

File: ipld/read.go [hunk 27]
```diff
@@ -18,190 +17,165 @@ import (
...
+}
 
-	// calculate the min number of shares needed to repair the square
-	minSharesNeeded := edsWidth * edsWidth / 4
+// fillQuadrant fetches 1/4 of shares for the given root
+func fillQuadrant(
+	ctx context.Context,
+	roots *quadrant, dag ipld.NodeGetter,
+	dataSquare [][]byte,
+) error {
```
Comment (renaynay) [inline ipld/read.go:100]:
```suggestion
	roots *quadrant, 
	dag ipld.NodeGetter,
	dataSquare [][]byte,
) error {
```

File: ipld/read.go [hunk 28]
```diff
@@ -18,190 +17,165 @@ import (
...
+	}
+	return quadrant, nil
+}
 
-	// calculate the min number of shares needed to repair the square
-	minSharesNeeded := edsWidth * edsWidth / 4
+// fillQuadrant fetches 1/4 of shares for the given root
+func fillQuadrant(
+	ctx context.Context,
+	roots *quadrant, dag ipld.NodeGetter,
```
Comment (renaynay) [inline ipld/read.go:98]:
```suggestion
	quadrant *quadrant, dag ipld.NodeGetter,
```

Comment (renaynay) [reply ipld/read.go:98]: let's just continue using this naming convention everywhere.

File: ipld/read.go (hunk 27 above)
Comment (renaynay) [reply ipld/read.go:100]: (pls also refactor uses of `roots` -> `quadrant` in this function

File: ipld/read.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (vgonkivs) [reply ipld/read.go:89]: Ok, I will create an issue.

File: ipld/read.go [hunk 29]
```diff
@@ -18,190 +17,165 @@ import (
...
+) error {
+	errGroup, ctx := errgroup.WithContext(ctx)
+	for i := 0; i < len(roots.rootCids); i++ {
+		i := i
+		errGroup.Go(func() error {
+			leaves, err := GetSubtreeLeaves(ctx, roots.rootCids[i], dag, roots.isLeftSubtree, uint32(len(roots.rootCids)/2))
+			if err != nil {
+				return err
+			}
+			length := len(roots.rootCids)
```
Comment (renaynay) [inline ipld/read.go:109]: could this be called `quadrantWidth` ?

Comment (renaynay) [reply ipld/read.go:109]: Since this is the "length" of all rootCIDs in the quadrant correct?

File: ipld/read.go [hunk 30]
```diff
@@ -18,190 +17,165 @@ import (
...
+			for leafIdx, leaf := range leaves {
+				// shares from the right subtree should be
+				// inserted after all shares from the left subtree
+				if !roots.isLeftSubtree {
+					leafIdx += len(leaves)
+				}
+				shareData := leaf.RawData()[1:]
+				dataSquare[((i+roots.from)*length*2)+leafIdx] = shareData[NamespaceSize:]
```
Comment (renaynay) [inline ipld/read.go:117]: Can you just document here how you calculate the position/index of where to add the share data inside the `dataSquare` ? Something like: 

```suggestion
                                 // position is calculated by < insert explanation>
                                 position := ((i+roots.from)*length*2)+leafIdx
				dataSquare[position] = shareData[NamespaceSize:]
```

**Review by renaynay — COMMENTED** [review]
small comments to improve readability IMO.

File: ipld/read.go (hunk 29 above)
Comment (vgonkivs) [reply ipld/read.go:109]: Yes

File: ipld/read.go (hunk 22 above)
Comment (vgonkivs) [reply ipld/read.go]: It is  used to calculate `position` in `dataSquare`

File: ipld/read.go (hunk 28 above)
Comment (vgonkivs) [reply ipld/read.go]: changed naming

File: ipld/read.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (vgonkivs) [reply ipld/read.go:89]: https://github.com/celestiaorg/celestia-node/issues/516

**Review by renaynay — APPROVED** [review]
all good on my side, waiting on @Wondertan

File: ipld/read.go [hunk 31]
```diff
@@ -18,190 +17,177 @@ import (
...
-	case sc.shareChan <- indexedShare{data: data[NamespaceSize:], index: index{row: rowIdx, col: colIdx}}:
-	}
+	return getLeaves(ctx, dag, subtreeRootHash, make([]ipld.Node, 0, treeSize))
 }
 
-// wait until enough data has been collected, the timeout has been reached, or
-// too many errors are encountered
-func (sc *shareCounter) wait() error {
-	defer sc.cancel()
+func GetLeaves(ctx context.Context, dag ipld.NodeGetter, root cid.Cid, size uint32) ([]ipld.Node, error) {
```
Comment (liamsi) [inline ipld/read.go:159]: Can this exported function be documented? Also can it have a unit test as well?

**Review by liamsi — COMMENTED** [review]
I think we should add tests and docs for the new exported functions. Otherwise looks good! Will take a closer look again after tests were added.

File: ipld/read.go (hunk 31 above)
Comment (vgonkivs) [reply ipld/read.go]: removed

File: ipld/read.go [hunk 32]
```diff
@@ -18,190 +17,177 @@ import (
...
 }
 
-// retrieveLeaf uses GetLeafData to fetch a single leaf and counts that leaf
-func (sc *shareCounter) retrieveShare(
-	rootCid cid.Cid,
-	isRow bool,
-	axisIdx uint32,
-	idx uint32,
+// GetSubtreeLeaves returns only one subtree - left or right
+func GetSubtreeLeaves(
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply ipld/read.go]: made as unexported

**Review by liamsi — APPROVED** [review]
👏🏼 🙏🏼
